cairo/crates/token/src/extensions/hyp_erc20_dex_collateral.cairo

```diff
@@ -0,0 +1,239 @@
+                ref self, amount_or_id,
+            )
+        }
+
+        fn transfer_to_hook(
+            ref self: TokenRouterComponent::ComponentState<ContractState>,
+            recipient: u256,
+            amount_or_id: u256,
+            metadata: Bytes,
+        ) {
+            let recipient: ContractAddress = recipient.try_into().unwrap();
+            let amount: felt252 = amount_or_id.try_into().unwrap();
+
+            let mut contract_state = TokenRouterComponent::HasComponent::get_contract_mut(ref self);
+            let dex_address = contract_state.dex.read();
+            let token_address = contract_state.collateral.wrapped_token.read().contract_address;
+
+            // Approve the DEX to spend the tokens
+            let token_dispatcher = ERC20ABIDispatcher { contract_address: token_address };
+            token_dispatcher.approve(dex_address, amount_or_id);
+
+            let mut calldata = ArrayTrait::new();
+            recipient.serialize(ref calldata); // the actual recipient of the deposit
+            token_address.serialize(ref calldata); // depositing collateral token
+            amount.serialize(ref calldata);
+
+            let dex_call_result = call_contract_syscall(
+                address: dex_address,
+                entry_point_selector: DEX_DEPOSIT_ON_BEHALF_OF_SELECTOR,
+                calldata: calldata.span(),
+            );
+            assert(dex_call_result.is_ok(), 'DEPOSIT_FAILED');
+
+            let mut dex_call_result_unwrapped = dex_call_result.unwrap();
+            let dex_call_success = Serde::<bool>::deserialize(ref dex_call_result_unwrapped)
+                .unwrap();
+            assert(dex_call_success, 'DEPOSIT_REJECTED');
+
+            contract_state
+                .emit(DexDeposit { token: token_address, recipient, amount: amount_or_id });
+        }
+    }
+
+
+    #[abi(embed_v0)]
+    impl HypErc20DexCollateralImpl of super::IHypErc20DexCollateral<ContractState> {
+        fn get_dex(self: @ContractState) -> starknet::ContractAddress {
+            self.dex.read()
+        }
+
+        fn get_deposit_token(self: @ContractState) -> starknet::ContractAddress {
+            self.collateral.wrapped_token.read().contract_address
+        }
+
+        fn balance_on_behalf_of(self: @ContractState, account: ContractAddress) -> u256 {
+            let dex_address = self.dex.read();
+            let token_address = self.collateral.wrapped_token.read().contract_address;
+
+            let mut calldata = ArrayTrait::new();
+            account.serialize(ref calldata);
+            token_address.serialize(ref calldata);
+
+            let balance_call_result = call_contract_syscall(
+                address: dex_address,
+                entry_point_selector: DEX_GET_TOKEN_ASSET_BALANCE_SELECTOR,
+                calldata: calldata.span(),
+            );
+            assert(balance_call_result.is_ok(), 'BALANCE_CALL_FAILED');
```

Comment:
ditto here, why not dispatcher?

Reply:
why are we implementing balanceOf this way if the user still has to withdraw from the exchange before `transferRemote`?

Reply:
I mimicked the call pattern from deposit which was recommended by Paradex.

This is to display balance on the warp UI.

Reply:
do we want our UI to display the paradex balance if the user still has to withdraw via their app?